lib/srv/db/opensearch/engine.go

```diff
@@ -0,0 +1,283 @@
+
+	awsSession, err := e.CloudClients.GetAWSSession(sessionCtx.Database.GetAWS().Region)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	svc, err := libaws.NewSigningService(libaws.SigningServiceConfig{
+		Clock:                 e.Clock,
+		Session:               awsSession,
+		GetSigningCredentials: e.GetSigningCredsFn,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	e.signingSvc = svc
+
+	return nil
+}
+
+// SendError sends an error to OpenSearch client.
+func (e *Engine) SendError(err error) {
+	if e.clientConn == nil || err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+
+	// TODO: ideally we would to switch over to strongly typed opensearch errors, but there is no equivalent of elastic.ErrorCause in opensearch libraries.
+	reason := err.Error()
+	cause := elastic.ErrorCause{
+		Reason: &reason,
+		Type:   "internal_server_error_exception",
+	}
+
+	// Assume internal server error HTTP 500 and override if possible.
+	statusCode := http.StatusInternalServerError
+	if trace.IsAccessDenied(err) {
+		statusCode = http.StatusUnauthorized
+		cause.Type = "access_denied_exception"
+	}
+
+	jsonBody, err := json.Marshal(cause)
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+
+	response := &http.Response{
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: statusCode,
+		Body:       io.NopCloser(bytes.NewBuffer(jsonBody)),
+		Header: map[string][]string{
+			"Content-Type":   {"application/json"},
+			"Content-Length": {strconv.Itoa(len(jsonBody))},
+		},
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.Errorf("OpenSearch error: %+v", trace.Unwrap(err))
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target OpenSearch server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	err := e.checkAccess(ctx)
+	e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// process reads request from connected OpenSearch client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, req *http.Request) error {
+	reqCopy, err := copyRequest(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// force HTTPS, set host URL.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = e.sessionCtx.Database.GetURI()
+
+	e.emitAuditEvent(reqCopy)
+
+	roleArn, err := libaws.BuildRoleARN(e.sessionCtx.DatabaseUser, e.sessionCtx.Database.GetAWS().Region, e.sessionCtx.Database.GetAWS().AccountID)
+	if err != nil {
+		return trace.Wrap(err)
```

Comment:
You can pass account id if you pass full ARN, not just the shortened one. I'll look at how we handle this elsewhere.

Reply:
All right, with the recent changes we require the account ID to be assigned at the database level. Whatever you pass as username must agree, but you can use shortened version (e.g. `--db-user=marek-role`) and it will work.